sources/org.osbuild.curl

```diff
@@ -83,93 +81,77 @@ SCHEMA = """
+            # rhsm secrets only need to be retrieved once and can then be reused
+            # Moving the subscriptions initialization here requires a lock. However it's possible to not pay the cost
+            # every time a reader wants to access the initialized version.
+            if self.subscriptions is None:
+                self.subscriptions = Subscriptions.from_host_system()
+                url["secrets"] = self.subscriptions.get_secrets(url.get("url"))
+        return checksum, url
+
+    # pylint: disable=arguments-renamed
+    # pylint: disable=too-many-branches
+    def fetch_one(self, checksum, url):
+        secrets = url.get("secrets")
+        url_path = url.get("url")
+        # Download to a temporary sub cache until we have verified the checksum. Use a
+        # subdirectory, so we avoid copying across block devices.
+        with tempfile.TemporaryDirectory(prefix="osbuild-unverified-file-", dir=self.cache) as tmpdir:
+            # some mirrors are sometimes broken. retry manually, because we could be
+            # redirected to a different, working, one on retry.
+            return_code = 0
+            for _ in range(10):
+                curl_command = [
+                    "curl",
+                    "--silent",
+                    "--speed-limit", "1000",
+                    "--connect-timeout", "30",
+                    "--fail",
+                    "--location",
+                    "--output", checksum,
+                ]
+                if secrets:
+                    if secrets.get('ssl_ca_cert'):
+                        curl_command.extend(["--cacert", secrets.get('ssl_ca_cert')])
+                    if secrets.get('ssl_client_cert'):
+                        curl_command.extend(["--cert", secrets.get('ssl_client_cert')])
+                    if secrets.get('ssl_client_key'):
+                        curl_command.extend(["--key", secrets.get('ssl_client_key')])
+                # url must follow options
+                curl_command.append(url_path)
+
+                curl = subprocess.run(curl_command, encoding="utf-8", cwd=tmpdir, check=False)
+                return_code = curl.returncode
+                if return_code == 0:
+                    break
+            else:
+                raise RuntimeError(f"curl: error downloading {url}: error code {return_code}")
+
+            if not verify_file(f"{tmpdir}/{checksum}", checksum):
+                raise RuntimeError(f"checksum mismatch: {checksum} {url}")
+
+            # The checksum has been verified, move the file into place. in case we race
+            # another download of the same file, we simply ignore the error as their
+            # contents are guaranteed to be  the same.
+            try:
+                os.rename(f"{tmpdir}/{checksum}", f"{self.cache}/{checksum}")
+            except FileExistsError:
+                pass
+
+    def config_download(self):
+        self.subscriptions = None
+        self.num_parallel_fetch = 4
```

Comment:
Both of these are: attribute definition outside of __init__ (pylint [W0201](https://pylint.pycqa.org/en/latest/messages/warning/attribute-defined-outside-init.html?highlight=attribute-defined-outside-init)).Attribute definition outside of __init__ (pylint [W0201](https://pylint.pycqa.org/en/latest/messages/warning/attribute-defined-outside-init.html?highlight=attribute-defined-outside-init)). But `subscriptions` can just be removed since it is only used in `transform`.

About `self.num_parallel_fetch`, I think this should be a class-property as `content_type`. Since in 99% of the cases we won't change this value per-instance (since it will only be `1`). We can set the default to `1` on `SourceService` and then adjust it only for the `CurlSource`.

Reply:
I agree for the class attribute. However, as I commented above for `subscriptions` it must remain an instance attribute since `transform` only work on one item at a time and we have to remember the state to not reinitialize them every time. I could however go for another design that does not uses an attribute but I'm not sure if it's worth it.